pkg/template/controller/templateinstance_controller.go

```diff
@@ -0,0 +1,174 @@
+import (
+	"net/http"
+
+	"github.com/golang/glog"
+	"github.com/openshift/origin/pkg/api/latest"
+	authclient "github.com/openshift/origin/pkg/auth/client"
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/config/cmd"
+	templateapi "github.com/openshift/origin/pkg/template/api"
+	templateapiv1 "github.com/openshift/origin/pkg/template/api/v1"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/meta/metatypes"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/auth/user"
+	"k8s.io/kubernetes/pkg/client/cache"
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	"k8s.io/kubernetes/pkg/client/restclient"
+	"k8s.io/kubernetes/pkg/kubectl/resource"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+type TemplateInstanceController struct {
+	restconfig restclient.Config
+	oc         *client.Client
+	controller *cache.Controller
+}
+
+func NewTemplateInstanceController(restconfig restclient.Config, oc *client.Client) *TemplateInstanceController {
+	c := TemplateInstanceController{restconfig: restconfig, oc: oc}
+
+	_, c.controller = cache.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options kapi.ListOptions) (runtime.Object, error) {
+				return c.oc.TemplateInstances(kapi.NamespaceAll).List(options)
+			},
+			WatchFunc: func(options kapi.ListOptions) (watch.Interface, error) {
+				return c.oc.TemplateInstances(kapi.NamespaceAll).Watch(options)
+			},
+		},
+		&templateapi.TemplateInstance{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.handle(obj.(*templateapi.TemplateInstance))
+			},
+			UpdateFunc: func(_, obj interface{}) {
+				c.handle(obj.(*templateapi.TemplateInstance))
+			},
+			DeleteFunc: func(obj interface{}) {
+			},
+		},
+	)
+
+	return &c
+}
+
+func (c *TemplateInstanceController) Run(stop <-chan struct{}) {
+	c.controller.Run(stop)
+}
+
+func (c *TemplateInstanceController) handle(templateInstance *templateapi.TemplateInstance) error {
+	for _, condition := range templateInstance.Status.Conditions {
+		if condition.Type == templateapi.TemplateInstanceCreated && condition.Status == kapi.ConditionTrue ||
+			condition.Type == templateapi.TemplateInstanceFailed && condition.Status == kapi.ConditionTrue {
+			return nil
+		}
+	}
+
+	err := c.provision(templateInstance)
+	if err == nil {
+		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
+			{
+				Type:               templateapi.TemplateInstanceCreated,
+				Status:             kapi.ConditionTrue,
+				LastTransitionTime: unversioned.Now(),
+				Reason:             "Created",
+			},
+		}
+
+	} else {
+		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
+			{
+				Type:               templateapi.TemplateInstanceFailed,
+				Status:             kapi.ConditionTrue,
+				LastTransitionTime: unversioned.Now(),
+				Reason:             "Failed",
```

Comment:
@smarterclayton it sounds like you're suggesting something more like phase that we're moving away from ("phase: failed" vs "phase: succeeded").

Assuming we're going to use conditions to represent terminal states, isn't it reasonable to have multiple mutually exclusive conditions that can be applied to an object?

Regardless, we should try to follow the pattern deploymentconfigs have outlined here...  assuming there is one.  Unfortunately I think DCs tend to get their state from the underlying RC rather than recording their own state.

conditions i would have expected:

new - request was created for the first time (could be represented by no conditions)
pending - controller has acknowledged and is provisioning resources (new condition gets removed?)
completed - resources were created (pending condition gets removed?)
failed - resources could not be created (pending condition gets removed?)
updated - something about it has changed, it will be re-provisioned (removes failed/complete condition).